File: 2023/day2/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
	"strconv"
	"strings"
)

func main() {
	file, err := os.Open(os.Args[1])
	if err != nil {
		panic(err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)

	resPart1 := 0
	resPart2 := 0

	for lineNumber := 1; scanner.Scan(); lineNumber++ {
		line := scanner.Text()

		if Part1(line) {
			resPart1 += lineNumber
		}

		powerNumber := Part2(line)
		resPart2 += powerNumber
	}

	fmt.Printf("Part_1: %d\n", resPart1)
	fmt.Printf("Part_2: %d\n", resPart2)
}

func Part1(input string) bool {
	game_subset := strings.Split((strings.Split(input, ":")[1]), ";")

	for _, val := range game_subset {

		pattern := regexp.MustCompile(`(\d+)\s+([a-zA-Z]+)`)
		matches := pattern.FindAllStringSubmatch(val, -1)

		for _, match := range matches {
			number, _ := strconv.Atoi(match[1])
			color := match[2]

			if (color == "red" && number > 12) ||
				(color == "green" && number > 13) ||
				(color == "blue" && number > 14) {
				return false
			}

		}
	}

	return true
}

func Part2(input string) int {
	game_subset := strings.Split((strings.Split(input, ":")[1]), ";")
	red := 0
	blue := 0
	green := 0
	for _, val := range game_subset {

		pattern := regexp.MustCompile(`(\d+)\s+([a-zA-Z]+)`)
		matches := pattern.FindAllStringSubmatch(val, -1)

		for _, match := range matches {
			number, _ := strconv.Atoi(match[1])
			color := match[2]
			if color == "red" && red < number {
				red = number
			}
			if color == "blue" && blue < number {
				blue = number
			}
			if color == "green" && green < number {
				green = number
			}

		}
	}

	return red * blue * green
}

File: 2023/day2/main.py
import sys
import re


def Part1(input_str):
    game_subset = input_str.split(":")[1].split(";")
    pattern = re.compile(r"(\d+)\s+([a-zA-Z]+)")
    for val in game_subset:
        matches = pattern.findall(val)
        for number, color in matches:
            number = int(number)
            if (
                (color == "red" and number > 12)
                or (color == "green" and number > 13)
                or (color == "blue" and number > 14)
            ):
                return False
    return True


def Part2(input_str):
    game_subset = input_str.split(":")[1].split(";")
    red, blue, green = 0, 0, 0
    pattern = re.compile(r"(\d+)\s+([a-zA-Z]+)")
    for val in game_subset:
        matches = pattern.findall(val)
        for number, color in matches:
            number = int(number)
            if color == "red":
                red = max(red, number)
            elif color == "blue":
                blue = max(blue, number)
            elif color == "green":
                green = max(green, number)
    return red * blue * green


if __name__ == "__main__":
    with open(sys.argv[1], "r") as file:
        res_part1 = 0
        res_part2 = 0
        for line_number, line in enumerate(file, 1):
            if Part1(line.strip()):
                res_part1 += line_number
            power_number = Part2(line.strip())
            res_part2 += power_number

    print(f"Part_1: {res_part1}")
    print(f"Part_2: {res_part2}")
